text/0000-immovable-types.md

```diff
@@ -20,47 +20,84 @@ Since generators can only move during suspend points we can require that referen
 # Detailed design
 [design]: #detailed-design
 
-A new unsafe auto trait `Move` is introduced in `core::marker`. Auto traits are implemented for all primitive types and for composite types where the elements also implement the trait. Users can opt-out to this for custom types. References, pointers, `core::ptr::Unique` and `core::ptr::Shared` explicitly implement this trait, since pointers are movable even if they point to immovable types.
-
-All type parameters (including `Self` for traits), trait objects and associated types have a `Move` bound by default.
+A new builtin marker trait `Move` is introduced in `core::marker`. All type parameters (including `Self` for traits) and associated types implement `Move` by default.
 
 If you want to allow types which may not implement `Move`, you would use the `?Move` trait bound which means that the type may or may not implement `Move`.
 
+A new marker struct `Immovable` is also introduced in `core::marker`. This struct does not implement `Move` and allows users to make composite immovable types.
+
 You can freely move values which are known to implement `Move` after they are borrowed, however you cannot move types which aren't known to implement `Move` after they have been borrowed. Once we borrow an immovable type, we'd know its address and code should be able to rely on the address not changing. This is sound since the only way to observe the address of a value is to borrow it. Before the first borrow nothing can observe the address and the value can be moved around.
 
 Static variables allow types which do not implement `Move`.
 
+These are the rules to determine if a type implements `Move`:
+- Integers, floats, `char` and `bool` are `Move`
+- Function types and function pointers  are `Move`
+- Closures are `Move` if their captured variables are `Move`
+- Movable generators are `Move` if all values (including capture variables) which are live during a suspension point are `Move`
+- Immovable generators are never `Move`
+- The `Immovable` type is never `Move`
+- Trait objects are `Move` if they have an explicit `Move` bound
+- Struct, enums and tuples are `Move` if all their elements are `Move`
+- Existential types (`impl Trait`) are `Move` if their underlying type are `Move`
```

Comment:
It might be worth mentioning here that (I assume?) `Move` "leaks" through `impl Trait` like other autotraits.